Delegate striping of debug symbols to GraalVM/Mandrel when >23.0

zakkak opened the pull request:
Starting with GraalVM/Mandrel 23.0, native-image strips debug info in a separate file by default.
    
Fixes #31258
Fixes https://github.com/Karm/mandrel-integration-tests/issues/138

cc @jerboaa

zakkak committed:
Simplify .debug file copying

jerboaa commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:
```diff
@@ -712,6 +712,15 @@ public NativeImageInvokerInfo build() {
+                    /*
+                     * Starting with GraalVM/Mandrel 23.0, native-image strips debug info in a separate file by default.
+                     * Quarkus already does this, so for the time being we disable GraalVM/Mandrel's splitting as in
+                     * contrast to Quarkus it uses absolute paths which makes debugging harder especially for
+                     * in-container builds.
```
I was too quick, sorry. There is no proof of this container issue. Even though the paths passed to `objcopy` are absolute, the links in the files end up being relative. So there seems to be no issue with upstream as far as I could tell. I'm not sure what the man page of `objcopy` alludes to, though. Anyway, perhaps we should use GraalVM's upstream version instead if it passes the failing integration tests?

zakkak replied:
Still there is another difference between the GraalVM and the Quarkus approach. The former strips all info (including symbols which it keeps in a different file though) from the image while Quarkus strips only debug info.

jerboaa replied:
That's right. I agree that it should be OK to do slightly different debuginfo handling between 22.3 and 23.x.

zakkak committed:
Delegate striping of debug symbols to GraalVM/Mandrel when >23.0

Starting with GraalVM/Mandrel 23.0, native-image strips debug info in a
separate file by default.

Fixes #31258

jerboaa commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildRunner.java:
```diff
@@ -66,21 +66,25 @@ public Result build(List<String> args, String nativeImageName, String resultingE
-                // Strip debug symbols regardless, because the underlying JDK might contain them
-                objcopy("--strip-debug", resultingExecutableName);
+                if (!debugSymbolsEnabled || graalVMVersion.compareTo(GraalVM.Version.VERSION_23_0_0) < 0) {
+                    // Strip debug symbols even if not generated by GraalVM/Mandrel, because the underlying JDK might
+                    // contain them
```
Could you please expand on the comment here? In particular we do not want to strip debuginfo on `>= GraalVM 23.x` when debug symbols are turned on, since in that case `native-image` itself would do the stripping.

jerboaa commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildRunner.java:
```diff
@@ -66,21 +66,25 @@ public Result build(List<String> args, String nativeImageName, String resultingE
+                    // contain them
+                    objcopy("--strip-debug", resultingExecutableName);
+                }
             } else if (SystemUtils.IS_OS_LINUX) {
```
So on a system without `objcopy` and GraalVM  `>= 23.0` and debuginfo turned on, we'd get duplicate warnings about `objcopy` missing. Once by quarkus a second time from `native-image`.

jerboaa reviewed (commented):
Looks good. Two minor things...

zakkak commented:
> Looks good. Two minor things...

@jerboaa I Added a new commit to address these two as well. Please have a look.

jerboaa reviewed (approved):
Looks fine to me. Thanks!

zakkak commented:
Once reviewed by someone else as well and the CI passes I will squash the commits and merge.

jerboaa commented:
@gsmet Could you please help get this reviewed and integrated? Thanks! This issue is causing some noise in our mandrel CI. I've verified the fix works, fwiw.

geoand commented:
We have a job failing due to one of the commits being a fixup commit. Let's address it before merging :)

jerboaa commented:
Thanks all!
